oxidation/libparsec/crates/platform_device_loader/src/testbed.rs

```diff
@@ -0,0 +1,89 @@
+// The idea would be as long as no api doing modification in local devices is
+// called, we can generate the LocalDevice object directly from template data
+// without having to do password derivation, deserialization and decryption.
+
+struct PseudoPersistentStorage {
+    available_devices: Vec<AvailableDevice>,
+}
+
+impl PseudoPersistentStorage {
+    pub fn new(
+        config_dir: &Path,
+        organization_addr: &BackendOrganizationAddr,
+        template: &TestbedTemplate,
+    ) -> Self {
+        // Populate the storage from the template.
+        // Once done we should no longer need the template data
+        let available_devices = template
+            .device_files
+            .iter()
+            .map(|device_file| {
+                let device = template
+                    .devices
+                    .iter()
+                    .find(|d| d.device_id == device_file.device_id)
+                    .unwrap();
+                let user = template
+                    .users
+                    .iter()
+                    .find(|u| &u.user_id == device_file.device_id.user_id())
+                    .unwrap();
+
+                AvailableDevice {
+                    key_file_path: config_dir
+                        .join(format!("{}.key", device_file.device_id.as_ref()))
+                        .into(),
+                    organization_id: organization_addr.organization_id().to_owned(),
+                    device_id: device_file.device_id.clone(),
+                    human_handle: user.human_handle.clone(),
+                    device_label: device.device_label.clone(),
+                    slug: local_device_slug(
+                        organization_addr.organization_id(),
+                        &device.device_id,
+                        organization_addr.root_verify_key(),
+                    ),
+                    ty: DeviceFileType::Password,
+                }
+            })
+            .collect();
+
+        Self { available_devices }
+    }
+}
+
+/// Retrieve (or create) our pseudo persistent storage, check it type and pass it to `cb`.
+/// Return `None` if `config_dir` doesn't correspond to a testbed env.
+fn with_pseudo_persistent_storage<T>(
+    config_dir: &Path,
+    cb: impl FnOnce(&mut PseudoPersistentStorage) -> T,
+) -> Option<T> {
+    test_get_testbed(config_dir).map(|env| {
+        let mut global_store = env.persistence_store.lock().expect("Mutex is poisoned");
+        let store =
+            global_store
+                .entry(STORE_ENTRY_KEY)
+                .or_insert(Box::new(PseudoPersistentStorage::new(
+                    config_dir,
+                    &env.organization_addr,
+                    env.template.as_ref(),
+                )));
+        let store = store
+            .downcast_mut::<PseudoPersistentStorage>()
+            .expect("Unexpected pseudo persistent storage type for platform_device_loader");
+        cb(store)
+    })
```

Comment:
Is the callback really needed ?
Why not directly return an `Option<PseudoPersistentStorage>` and use `map` on other functions ?

Reply:
The function taking a callback is needed because we need to keep the lock on `env.persistence_store` for the whole time the callback is called.